core/src/main/java/org/jdbi/v3/ConstructorMapper.java

```diff
@@ -0,0 +1,174 @@
+
+/**
+ * A row mapper which maps the fields in a result set into a constructor. The default implementation will perform a
+ * case insensitive mapping between the constructor parameter names and the column labels,
+ * also considering camel-case to underscores conversion.
+ *
+ * Currently the constructor must have exactly the same number of columns as the result set, and
+ * the mapping must be one-to-one.  These restrictions may be reconsidered at a later time.
+ */
+public class ConstructorMapper<T> implements RowMapper<T>
+{
+    private final Constructor<T> constructor;
+    private final ConcurrentMap<List<String>, ObjectCreator<T>> creatorCache = new ConcurrentHashMap<>();
+    private final Collection<ColumnNameMappingStrategy> nameMappingStrategies;
+
+    private ConstructorMapper(Constructor<T> constructor, Collection<ColumnNameMappingStrategy> nameMappingStrategies)
+    {
+        this.constructor = constructor;
+        this.nameMappingStrategies = Collections.unmodifiableList(new ArrayList<>(nameMappingStrategies));
+    }
+
+    @Override
+    public T map(ResultSet rs, StatementContext ctx)
+        throws SQLException
+    {
+        final ResultSetMetaData metadata = rs.getMetaData();
+        final List<String> columnNames = new ArrayList<>(metadata.getColumnCount());
+
+        for (int i = 1; i <= metadata.getColumnCount(); ++i) {
+            columnNames.add(metadata.getColumnLabel(i));
+        }
+
+        ObjectCreator<T> creator = creatorCache.computeIfAbsent(columnNames, c -> createCreator(c, metadata, ctx));
+        return creator.create(rs);
+    }
+
+    private ObjectCreator<T> createCreator(List<String> columnNames, ResultSetMetaData metadata, StatementContext ctx) {
+        final int length;
+        try {
+            length = metadata.getColumnCount();
+        } catch (SQLException e) {
+            throw new ResultSetException("Could not get column count", e, ctx);
+        }
+
+        if (length != constructor.getParameterCount()) {
+            throw new IllegalStateException(length + " columns in result set, but constructor takes " + constructor.getParameterCount());
+        }
+
+        final int[] columnMap = new int[length];
+        final ColumnMapper<?>[] mappers = new ColumnMapper<?>[length];
+
+        for (int i = 0; i < length; i++) {
+            final Type type = constructor.getGenericParameterTypes()[i];
+            final int paramIndex = parameterIndexForColumn(columnNames.get(i));
+
+            if (mappers[paramIndex] != null) {
+                throw new IllegalArgumentException("Column named " + columnNames.get(i) + " maps to multiple parameters");
+            }
+
+            mappers[paramIndex] = ctx.findColumnMapperFor(type).orElseThrow(() ->
+                new IllegalArgumentException("Could not find column mapper for type '" + type + "' of parameter for constructor " + constructor));
+            columnMap[i] = paramIndex;
+        }
```

Comment:
I we go with the column name to parameter name mapping, it would probably make sense to have a sanity check after this loop, to verify each column is used to exactly one parameter.